crc32.c

```diff
@@ -217,22 +282,31 @@ unsigned long ZEXPORT crc32_z(crc, buf, len)
 
         endian = 1;
         if (*((unsigned char *)(&endian)))
-            return crc32_little(crc, buf, len);
+            return crc32_little;
         else
-            return crc32_big(crc, buf, len);
+            return crc32_big;
     }
 #endif /* BYFOUR */
-    crc = crc ^ 0xffffffffUL;
-    while (len >= 8) {
-        DO8;
-        len -= 8;
-    }
-    if (len) do {
-        DO1;
-    } while (--len);
-    return crc ^ 0xffffffffUL;
+
+    return crc32_table_lookup;
 }
 
+#if !defined(Z_IFUNC_ASM) && !defined(Z_IFUNC_NATIVE)
+
+unsigned long ZEXPORT crc32_z(crc, buf, len)
+    unsigned long crc;
+    const unsigned char FAR *buf;
+    z_size_t len;
+{
+    static unsigned long ZEXPORT (*crc32_func)(unsigned long, const unsigned char FAR *, z_size_t) = NULL;
```

Comment:
Adding a static plus invoking a function pointer seems potentially more expensive than testing for a bool (at least on ARM it was).

I'm talking about:
https://chromium-review.googlesource.com/c/chromium/src/+/801108/12/third_party/zlib/crc32.c

Reply:
Have you got a test framework you tried on this? I'm keen to make something efficient across all platforms. I'm trusting the glibc folks a lot to get IFUNC right for the shard library case  as the crc32_z_ifunc return pointer becomes crc32_z in the trampolines and wouldn't mind to validate efficiencies of function pointer vs bool vs ifunc ways.

Reply:
Our optimization work in zlib targeted to improve PNG decoding performance in Chromium (that has both tracing tools as also an image_decoder benchmarker).

The results using a function pointer were worst in the device tested (ARMv8) for the PNG corpus we used (but remember that Chromium is built for ARM using clang).

Also Chromium has just added a compression/decompression benchmarker, it may be helpful to have a look on:
https://cs.chromium.org/chromium/src/third_party/zlib/contrib/bench/